### PR TITLE
Plugin duplicate protection (identifiers and directory names)

### LINKED ISSUE #223: Prevent duplicate plugin submissions by plugin identifier before zipfile is accepted for submission
Duplicate protection by plugin directory name is handled in #212, ie a user cannot submit `/models/my_model` if a directory `my_model` already exists in the `models` directory (UNLESS they are the original submitter of that directory).

We ALSO need to ensure that we are handling duplicate protection for plugin identifiers, ie a user cannot submit a benchmark with the identifier `benchmark_identifier` if a benchmark with that identifier already exists. Currently, PR checks will fail, so while the submission won't make it to automerge, catching it at this later point will be frustrating to the user and should be handled before the zipfile is accepted by the website. This is a bit trickier than directory name duplication protection as it requires parsing all existing benchmark registrations.

### PR DESCRIPTION
**Pull request by samwinebrake**
- added extract_identifiers to find patterns within plugin inits
- changed variable naming within is_submission_original
- duplicate protection of plugin identifiers within inits

@mike-ferguson and @shehadak will this work with new changes being pushed shortly?

Closes #223

### CODE REVIEW

File: benchmarks/views/user.py
Comment (shehadak) [inline benchmarks/views/user.py:338]: Won't these also match commented lines? E.g. # model_registry["alexnet"] = .... I would update this to only include non-commented lines.

Comment (samwinebrake) [reply benchmarks/views/user.py:338]: Good catch! Should be fixed now so both inline and own line comments are ignored.

File: benchmarks/views/user.py
Comment (mike-ferguson) [inline benchmarks/views/user.py:216]: Why the removal of this return type hint?

**Review by mike-ferguson — COMMENTED** [review]
LGTM - just one comment, see attached

**Comment by mike-ferguson** [comment]
Also: #224 Is almost merged, once tests pass. I say hold off on merging this PR and let's fold in those changes from #224 into here and do a quick test. I don't think anything will conflict, but it will be good to check

File: benchmarks/views/user.py
Comment (shehadak) [inline benchmarks/views/user.py:348]: Consider a case (which already exists) of a submission that looks like this:
```
models/
    some_model/
        utils/
            __init__.py
            some_utils_file.py
         __init__.py
         ...
```
It is possible that traversing through `zip_ref.infolist()` will pass through `models/some_model/utils/__init__.py` before `models/some_model/__init__.py`, in which case the utils `__init__` will be used to look for plugin registrations, not the `some_model/__init__.py`. Therefore, I don't think the two conditions that `plugin in path_Segments` and `'__init__.py` in path_segments[-1]` are sufficient. I hope this failure mode makes sense.

**Comment by shehadak** [comment]
LGTM as well once comments are resolved! :-)

File: benchmarks/views/user.py
Comment (samwinebrake) [reply benchmarks/views/user.py]: fixed so only 'zip_root/plugin/plugin_name/__init__.py' inits are supported

**Review by shehadak — APPROVED** [review]

**Pull request merged**
